sql/binlog.cc

```diff
@@ -7631,36 +7631,43 @@ uint MYSQL_BIN_LOG::next_file_id() {
 
 extern "C" char mysql_bin_log_is_open(void) { return mysql_bin_log.is_open(); }
 
-extern "C" void mysql_bin_log_lock_commits(void) {
-  mysql_bin_log.lock_commits();
+extern "C" void mysql_bin_log_lock_commits(struct snapshot_info_st *ss_info) {
+  mysql_bin_log.lock_commits(ss_info);
 }
 
-extern "C" void mysql_bin_log_unlock_commits(char *binlog_file,
-                                             unsigned long long *binlog_pos,
-                                             char **gtid_executed,
-                                             int *gtid_executed_length,
-                                             unsigned long long *snapshot_hlc) {
-  mysql_bin_log.unlock_commits(binlog_file, binlog_pos, gtid_executed,
-                               gtid_executed_length, snapshot_hlc);
+extern "C" void mysql_bin_log_unlock_commits(struct snapshot_info_st *ss_info) {
+  mysql_bin_log.unlock_commits(ss_info);
 }
 
-void MYSQL_BIN_LOG::lock_commits(void) {
+void MYSQL_BIN_LOG::lock_commits(snapshot_info_st *ss_info) {
   mysql_mutex_lock(&LOCK_log);
   drain_committing_trxs(true);
+
+  ss_info->binlog_file = std::string(log_file_name);
+  ss_info->binlog_pos = my_b_tell(&index_file);
+
+  global_sid_lock->wrlock();
+  char *gtid_buff = nullptr;
+  gtid_state->get_executed_gtids()->to_string(&gtid_buff);
+  ss_info->gtid_executed = std::string(gtid_buff);
```

Comment:
`char*` can be assigned to std::string without temporary object
```suggestion
  ss_info->gtid_executed = gtid_buff;
```
Also add protection against `gtid_buff == nullptr`